README.rst
django-mediawiki-auth
=====================

Allows django apps to have seamless log-in sessions with MediaWiki.

Built for `Open Source Ecology <http://opensourceecology.org>`_


CHANGES.rst
0.0.2
-----

+ Added admin.py

0.0.1
-----

* Initial release.


mediawiki_auth/tests/urls.py
from django.conf.urls import url
from .views import *

urlpatterns = [
    url(r'^$', Status.as_view(), name='status'),
]


mediawiki_auth/tests/utils.py
import os; os.environ['DJANGO_SETTINGS_MODULE'] = 'mediawiki_auth.tests.settings'
import django; django.setup()
from django.contrib.staticfiles.testing import StaticLiveServerTestCase

from selenium import webdriver
from selenium.webdriver.chrome.options import Options


class ChromeTestCase(StaticLiveServerTestCase):

    port = 8081

    @classmethod
    def setUpClass(cls):
        super().setUpClass()
        options = Options()
        options.add_argument("--headless")
        cls.browser = webdriver.Chrome(chrome_options=options)

    @classmethod
    def tearDownClass(cls):
        cls.browser.quit()
        super().tearDownClass()

    def setUp(self):
        super().setUp()
        self.browser.delete_all_cookies()

    def login_into_wiki(self):
        self.browser.get('http://wiki.test.localhost/index.php?title=Special:UserLogin')
        self.browser.find_element_by_name('wpName').send_keys('admin')
        self.browser.find_element_by_name('wpPassword').send_keys('a-very-long-password')
        self.browser.find_element_by_name('wpLoginAttempt').click()

    def logout_of_wiki(self):
        self.browser.get('http://wiki.test.localhost/index.php?title=Special:UserLogout')


mediawiki_auth/tests/views.py
from django.views.generic import View
from django.http import HttpResponse


class Status(View):

    def get(self, request, *args, **kwargs):
        if request.user.is_authenticated:
            return HttpResponse('<b id="status">logged-in as {u.username}</b>'.format(u=request.user))
        else:
            return HttpResponse('<b id="status">not logged-in</b>')


mediawiki_auth/admin.py
from django.contrib import admin
from .models import MediaWikiUser
from .mediawiki import (
    get_django_user_from_mediawiki_user,
    get_or_create_django_user_from_mediawiki_user
)


@admin.register(MediaWikiUser)
class MediaWikiUserAdmin(admin.ModelAdmin):
    list_per_page = 20
    list_display = (
        'name', 'real_name', 'email', 'last_accessed', 'registered', 'edit_count', 'linked'
    )
    search_fields = (
        'name', 'real_name', 'email'
    )
    fields = readonly_fields = ('id',)+list_display

    def get_actions(self, request):
        return {'link': (MediaWikiUserAdmin.link, 'link', "Link selected MediaWiki users into django")}

    def has_add_permission(self, request):
        return False

    def has_delete_permission(self, request, obj=None):
        return False

    def link(self, request, queryset):
        users = []
        for user in queryset:
            _, created = get_or_create_django_user_from_mediawiki_user(user)
            if created:
                users.append("'{}'".format(user.name))
        self.message_user(request, "%s successfully linked.".format(', '.join(users)))

    def linked(self, obj):
        return bool(get_django_user_from_mediawiki_user(obj))
    linked.boolean = True
    linked.short_description = 'Linked'


mediawiki_auth/models.py
from datetime import datetime
from django.db import models
from django.conf import settings
from django.contrib.auth.models import UserManager


MEDIAWIKI_DB_ALIAS = getattr(settings, 'MEDIAWIKI_DB_ALIAS', 'mediawiki')
MEDIAWIKI_TABLE_PREFIX = getattr(settings, 'MEDIAWIKI_TABLE_PREFIX', '')
MEDIAWIKI_USER_FIELD = getattr(settings, 'MEDIAWIKI_USER_FIELD', 'mediawiki_user')
MEDIAWIKI_USER_FK_FIELD = getattr(settings, 'MEDIAWIKI_USER_FK_FIELD', 'mediawiki_user_id')


def ts_to_dt(ts):
    if len(ts) == 14:
        return datetime(
            int(ts[:4]),
            int(ts[4:6]),
            int(ts[6:8]),
            int(ts[8:10]),
            int(ts[10:12]),
            int(ts[12:]),
        )


class MediaWikiEnabledUserManager(UserManager):

    def create_from_mediawiki_user(self, wiki_user):
        email = self.normalize_email(wiki_user.email)
        username = self.model.normalize_username(wiki_user.name)
        first, last = self.normalize_first_last(wiki_user)
        user = self.model(
            username=username,
            email=email,
            first_name=first,
            last_name=last,
            date_joined=wiki_user.registered
        )
        setattr(user, MEDIAWIKI_TABLE_PREFIX, wiki_user)
        setattr(user, MEDIAWIKI_USER_FK_FIELD, wiki_user.id)
        user.save(using=self._db)
        return user

    @staticmethod
    def normalize_first_last(wiki_user):
        first, last = '', ''
        if wiki_user.real_name:
            parts = wiki_user.real_name.split(' ')
            first = parts[0]
            if len(parts) > 1:
                last = ' '.join(parts[1:])
        return first, last


class MediaWikiUserManager(models.Manager):

    def get_queryset(self):
        return super().get_queryset().using(MEDIAWIKI_DB_ALIAS)


class MediaWikiUser(models.Model):

    class Meta:
        verbose_name = "MediaWiki User"
        verbose_name_plural = "MediaWiki Users"
        ordering = ('name',)
        db_table = MEDIAWIKI_TABLE_PREFIX+'user'
        managed = False

    id = models.AutoField(primary_key=True, db_column='user_id')
    name = models.BinaryField(max_length=255, db_column='user_name')
    real_name = models.CharField(max_length=255, db_column='user_real_name')
    email = models.CharField(max_length=255, db_column='user_email')

    token = models.BinaryField(max_length=32, db_column='user_token')

    email_authenticated = models.BinaryField(max_length=14, db_column='user_email_authenticated')
    email_token = models.BinaryField(max_length=32, db_column='user_email_token')
    email_token_expires = models.BinaryField(max_length=14, db_column='user_email_token_expires')

    registration = models.BinaryField(max_length=14, db_column='user_registration')
    touched = models.BinaryField(max_length=14, db_column='user_touched')

    edit_count = models.PositiveIntegerField(db_column='user_editcount')

    objects = MediaWikiUserManager()

    def verify_session_and_cookie_values(self, session, user_id):
        return self.id == session[b'wsUserID'] == int(user_id)

    @property
    def registered(self):
        return ts_to_dt(self.registration)

    @property
    def last_accessed(self):
        return ts_to_dt(self.touched)


class RecentChangesManager(models.Manager):

    def get_queryset(self):
        return super().get_queryset().using(MEDIAWIKI_DB_ALIAS)


class RecentChanges(models.Model):

    class Meta:
        db_table = MEDIAWIKI_TABLE_PREFIX+'recentchanges'
        managed = False
        ordering = ('-rc_timestamp',)

    id = models.AutoField(primary_key=True, db_column='rc_id')
    title = models.CharField(max_length=255, db_column='rc_title')
    user = models.ForeignKey(MediaWikiUser, db_column='rc_user')
    username = models.CharField(max_length=255, db_column='rc_user_text')
    rc_timestamp = models.BinaryField()

    objects = RecentChangesManager()

    @property
    def modified(self):
        return ts_to_dt(self.rc_timestamp)


class RevisionManager(models.Manager):

    def get_queryset(self):
        return super().get_queryset().using(MEDIAWIKI_DB_ALIAS)


class Revision(models.Model):

    class Meta:
        db_table = MEDIAWIKI_TABLE_PREFIX+'revision'
        managed = False

    id = models.AutoField(primary_key=True, db_column='rev_id')
    rev_timestamp = models.BinaryField()

    objects = RevisionManager()

    @property
    def modified(self):
        return ts_to_dt(self.rev_timestamp)


mediawiki_auth/mediawiki.py
import re
import zlib
import phpserialize
from django.db import connections
from django.conf import settings
from django.contrib.auth import get_user as get_django_user_from_request
from django.contrib.auth import get_user_model as get_django_user_model
from django.contrib.auth import login

from .models import (
    MediaWikiUser,
    MEDIAWIKI_USER_FK_FIELD,
    MEDIAWIKI_USER_FIELD,
    MEDIAWIKI_TABLE_PREFIX,
    MEDIAWIKI_DB_ALIAS
)

MEDIAWIKI_COOKIE_PREFIX = getattr(settings, 'MEDIAWIKI_COOKIE_PREFIX', 'mediawiki')


def unserialize_session(val):
    session = {}
    groups = re.split(b"([a-zA-Z0-9_]+)\|", val)
    if len(groups) > 2:
        groups = groups[1:]
        groups = zip(*([iter(groups)] * 2))
    for key, php_value in groups:
        session[key] = phpserialize.loads(php_value)
    return session


def get_session(session_id):
    with connections[MEDIAWIKI_DB_ALIAS].cursor() as c:
        c.execute("select value from {}objectcache where keyname = %s".format(MEDIAWIKI_TABLE_PREFIX), (
            'mediawiki:session:'+session_id,
        ))
        result = c.fetchone()
        if result and result[0]:
            serialized = zlib.decompress(result[0], -15)
            return unserialize_session(serialized)


def get_session_from_request(request):
    session_id = request.COOKIES.get(MEDIAWIKI_COOKIE_PREFIX+'_session')
    if session_id:
        return get_session(session_id)


def get_user(user_id):
    try:
        return MediaWikiUser.objects.get(pk=user_id)
    except (MediaWikiUser.DoesNotExist, MediaWikiUser.MultipleObjectsReturned):
        return None


def get_mediawiki_user_from_request(request):
    session = get_session_from_request(request)
    if session and session.get(b'wsUserID'):
        user = get_user(session[b'wsUserID'])
        if user:
            user_id = request.COOKIES.get(MEDIAWIKI_COOKIE_PREFIX+'UserID')
            if user.verify_session_and_cookie_values(session, user_id):
                return user


def get_django_user_from_mediawiki_user(wiki_user):
    User = get_django_user_model()
    try:
        return User.objects.get(**{MEDIAWIKI_USER_FK_FIELD: wiki_user.id})
    except (User.DoesNotExist, User.MultipleObjectsReturned):
        return None


def get_or_create_django_user_from_mediawiki_user(wiki_user):
    DjangoUser = get_django_user_model()
    try:
        return DjangoUser.objects.get(**{MEDIAWIKI_USER_FK_FIELD: wiki_user.id}), False
    except DjangoUser.DoesNotExist:
        return DjangoUser.objects.create_from_mediawiki_user(wiki_user), True


def get_or_create_django_user(request):
    from django.contrib.auth.models import AnonymousUser

    wiki_user = get_mediawiki_user_from_request(request)
    if wiki_user:

        django_user = get_django_user_from_request(request)
        if django_user.is_authenticated:

            if wiki_user.id == getattr(django_user, MEDIAWIKI_USER_FK_FIELD):
                # wiki session and django sessions match
                setattr(django_user, MEDIAWIKI_USER_FIELD, wiki_user)
                return django_user
            else:
                # sessions mismatch, invalidate django session
                request.session.flush()

        else:

            django_user, _ = get_or_create_django_user_from_mediawiki_user(wiki_user)
            login(request, django_user)
            setattr(django_user, MEDIAWIKI_USER_FIELD, wiki_user)
            return django_user

    return AnonymousUser()


mediawiki_auth/middleware.py
from django.conf import settings
from django.utils.deprecation import MiddlewareMixin
from django.utils.functional import SimpleLazyObject
from mediawiki_auth import mediawiki


def get_user(request):
    if not hasattr(request, '_cached_user'):
        request._cached_user = mediawiki.get_or_create_django_user(request)
    return request._cached_user


class AuthenticationMiddleware(MiddlewareMixin):
    def process_request(self, request):
        assert hasattr(request, 'session'), (
            "The Django authentication middleware requires session middleware "
            "to be installed. Edit your MIDDLEWARE%s setting to insert "
            "'django.contrib.sessions.middleware.SessionMiddleware' before "
            "'django.contrib.auth.middleware.AuthenticationMiddleware'."
        ) % ("_CLASSES" if settings.MIDDLEWARE is None else "")
        request.user = SimpleLazyObject(lambda: get_user(request))


mediawiki_auth/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.11.3 on 2017-07-23 12:44
from __future__ import unicode_literals

import django.contrib.auth.validators
from django.db import migrations, models
import django.utils.timezone
import mediawiki_auth.models


class Migration(migrations.Migration):

    initial = True

    dependencies = [
        ('auth', '0008_alter_user_username_max_length'),
    ]

    operations = [
        migrations.CreateModel(
            name='MediaWikiUser',
            fields=[
                ('id', models.AutoField(db_column='user_id', primary_key=True, serialize=False)),
                ('name', models.BinaryField(db_column='user_name', max_length=255)),
                ('real_name', models.CharField(db_column='user_real_name', max_length=255)),
                ('email', models.CharField(db_column='user_email', max_length=255)),
                ('touched', models.BinaryField(db_column='user_touched', max_length=14)),
                ('token', models.BinaryField(db_column='user_token', max_length=32)),
                ('email_authenticated', models.BinaryField(db_column='user_email_authenticated', max_length=14)),
                ('email_token', models.BinaryField(db_column='user_email_token', max_length=32)),
                ('email_token_expires', models.BinaryField(db_column='user_email_token_expires', max_length=14)),
                ('registration', models.BinaryField(db_column='user_registration', max_length=14)),
                ('editcount', models.PositiveIntegerField(db_column='user_editcount')),
            ],
            options={
                'managed': False,
                'db_table': 'user',
            },
        ),
        migrations.CreateModel(
            name='DjangoUser',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('password', models.CharField(max_length=128, verbose_name='password')),
                ('last_login', models.DateTimeField(blank=True, null=True, verbose_name='last login')),
                ('is_superuser', models.BooleanField(default=False, help_text='Designates that this user has all permissions without explicitly assigning them.', verbose_name='superuser status')),
                ('username', models.CharField(error_messages={'unique': 'A user with that username already exists.'}, help_text='Required. 150 characters or fewer. Letters, digits and @/./+/-/_ only.', max_length=150, unique=True, validators=[django.contrib.auth.validators.UnicodeUsernameValidator()], verbose_name='username')),
                ('first_name', models.CharField(blank=True, max_length=30, verbose_name='first name')),
                ('last_name', models.CharField(blank=True, max_length=30, verbose_name='last name')),
                ('email', models.EmailField(blank=True, max_length=254, verbose_name='email address')),
                ('is_staff', models.BooleanField(default=False, help_text='Designates whether the user can log into this admin site.', verbose_name='staff status')),
                ('is_active', models.BooleanField(default=True, help_text='Designates whether this user should be treated as active. Unselect this instead of deleting accounts.', verbose_name='active')),
                ('date_joined', models.DateTimeField(default=django.utils.timezone.now, verbose_name='date joined')),
                ('mediawiki_user_id', models.PositiveIntegerField(db_index=True, unique=True)),
                ('groups', models.ManyToManyField(blank=True, help_text='The groups this user belongs to. A user will get all permissions granted to each of their groups.', related_name='user_set', related_query_name='user', to='auth.Group', verbose_name='groups')),
                ('user_permissions', models.ManyToManyField(blank=True, help_text='Specific permissions for this user.', related_name='user_set', related_query_name='user', to='auth.Permission', verbose_name='user permissions')),
            ],
            options={
                'verbose_name_plural': 'users',
                'abstract': False,
                'verbose_name': 'user',
            },
            managers=[
                ('objects', mediawiki_auth.models.DjangoUserManager()),
            ],
        ),
    ]
